sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -794,6 +795,108 @@ class SparkSession private(
     }
   }
 
+
+  /**
+   * Add a tag to be assigned to all the operations started by this thread in this session.
+   *
+   * Often, a unit of execution in an application consists of multiple Spark executions.
+   * Application programmers can use this method to group all those jobs together and give a group
+   * tag. The application can use `org.apache.spark.sql.SparkSession.interruptTag` to cancel all
+   * running executions with this tag. For example:
+   * {{{
+   * // In the main thread:
+   * spark.addTag("myjobs")
+   * spark.range(10).map(i => { Thread.sleep(10); i }).collect()
+   *
+   * // In a separate thread:
+   * spark.interruptTag("myjobs")
+   * }}}
+   *
+   * There may be multiple tags present at the same time, so different parts of application may
+   * use different tags to perform cancellation at different levels of granularity.
+   *
+   * @param tag
+   *   The tag to be added. Cannot contain ',' (comma) character or be an empty string.
+   *
+   * @since 4.0.0
+   */
+  def addTag(tag: String): Unit = sparkContext.addJobTag(tag)
```

Comment:
This seems to make the strong assumption that a session is only used in a single thread. That does not need to hold.

Reply:
We may have to take a small step back here. I think we have to track the tags in the session it self, and only set them when we execute an action.

Reply:
I think we also need to make sure the tag is unique to this session.

Reply:
Done. We now store all user-provided tags in the session, and on the backend, we actually attach `spark-$sessionUUID-$tag` to jobs started by this session.